### ISSUE TITLE
memory leak in connect / disconnect

### ISSUE THREAD

**Issue by or4cle** [rana/ora#148]
Hello,

I am new to Go, and have been writing a Prometheus exporter for Oracle db. Basically it lets users specify arbitrary queries, and then transforms the result set to Prometheus metrics.

I believe I have stumbled upon a memory leak that looks very similar to what was reported in [this issue](https://github.com/rana/ora/issues/69) 

The foundation I used for the Oracle part originated from this: [csvdump.go](https://github.com/tgulacsi/go/blob/master/dbcsv/csvdump/csvdump.go)

In fact, if we modify csvdump.go slightly to loop over and over, we can reproduce the memory leak. 

Here is my test case:
```
create table foo(label1 varchar2(32), label2 varchar2(32), label3 varchar2(32), label4 varchar2(32), v number);

insert into foo select sys_guid(), sys_guid(), sys_guid(), sys_guid(), dbms_random.random from dual connect by level <= 10000;

commit;
```

A modified version of the csvdumper that loops forever is [here](http://pastebin.com/kSkdHk0t)

Once compiled, we can run:
`./csvdump -connect user/pass@db:port/sid foo > /dev/null`

If you let this run for a few hours, you'll see the process RSS keeps on increasing. I read over the memory leak described in issue 69, but am not sure I understand what exactly was done to fix it. Given this csvdump implementation, is there an easy way to change how we interface with Scan() to avoid the leak?

**Comment by tgulacsi** [comment]
First, create the db connectio before the loop, and close it with defer -
database/sql.DB maintains its own connection pool, so this close/reopen is
unneeded.

If the memory still grows, the I'll have to look into it more thoroughly...

or4cle <notifications@github.com> ezt írta (időpont: 2017. jan. 16., H
23:57):

**Comment by or4cle** [comment]
Hi,

Thanks for the quick reply!

You are correct, once I moved the database connect/dis-connect out of the loop, memory consumption dropped substantially. It is still increasing, but much less than before. For the previous 10 hours, grabbing 100k rows every 2 seconds (using the table defined above) RSS increases by 3M (whereas previously it would have been hundreds of megs). I will keep monitoring it, as I'm curious if it will stabilize at some point.

I wonder if this means we are leaking somewhere in the db open/close (perhaps oci handles)? I realize that in many situations, optimally, the connection should be maintained / persistent. But I also think some reasonable amount of connect/disconnect (say connecting to run one query every few minutes) should not cause the heap to grow unbounded. What do you think?

**Comment by or4cle** [comment]
Anecdotally, back in the day (perhaps ten years ago), I did some OCI programming and faced similar issues. I wanted to be able to loop the code a bazillion times and not have it leak. From what I remember, I had to track all OCIHandleAlloc()s and OCIHandleFree() them explicitly. 

The Oracle docs are a little hazy on this aspect, e.g.: [OCIHandleFree() doc](http://docs.oracle.com/database/121/LNOCI/oci16rel002.htm#LNOCI17135)

Mentions "All handles may be explicitly deallocated", but then goes on to say "The OCI deallocates a child handle if the parent is deallocated". However, in my estimation, it never clearly defines which handles are considered child handles. For example, this:

http://docs.oracle.com/database/121/LNOCI/oci02bas.htm#sthref125

only talks about statements and binds, but does not define any relationship among other handles (e.g. env and error).

Additionally, if we look at some header files included in the rdbms distribution, e.g. $ORACLE_HOME/rdbms/public/ociap.h, we see:

> OCIHandleFree()
> Name
> OCI Free HaNDLe
> Purpose
> This call explicitly deallocates a handle.
> Syntax
> sword OCIHandleFree ( void     *hndlp,
>                     ub4       type);
> Comments
> This call frees up storage associated with a handle, corresponding to the type
> specified in the type parameter.
> This call returns either OCI_SUCCESS or OCI_INVALID_HANDLE.
> All handles must be explicitly deallocated. OCI will not deallocate a child
> handle if the parent is deallocated.
> 
> 

There it is telling us exactly the opposite!

Furthermore, from their sample OCI code I've reviewed online, they always seem to explicitly deallocate everything. 

In any case, I went ahead and created a fork/branch and updated the code to explicitly deallocate the OCI_HTYPE_* handles. Once I made this change and ran the original csvdump code (that connects and disconnects repeatedly), I no longer see the same rate of uncontrolled heap growth. I cannot say it has zero leaks, but it's definitely way less than before. I'll submit a pull request so you can merge/use it as you see fit.

**Comment by tgulacsi** [comment]
Thanks!
I've merged this into both v3 and v4 versions, but had to remove the freeing of the error handler, as that resulted in SIGSEGV.

**Comment by or4cle** [comment]
Hi,

Could you let me know how to reproduce the SIGSEV, and I can look into it? 

I tried running "go test" under gopkg.in/rana/ora.v4, however I get this failure without any changes:

```
--- FAIL: TestPool (1.55s)
        z_server_session_test.go:139: After work: (115,62) -> (115,62)
        z_server_session_test.go:139: Pool close: (115,62) -> (119,64)
        z_server_session_test.go:141: Pool close process count went up 9!
        z_server_session_test.go:144: Pool close session count went up 19!
--- FAIL: ExampleDrvStmt_Exec_Query (0.10s)
```

**Comment by tgulacsi** [comment]
$ fgrep -n SIGSEGV *.go
env.go:124:     // [signal SIGSEGV: segmentation violation code=0x1
addr=0xbfc908 pc=0x7f65bf5ce0d6]

Se uncomment the freeOciHandle on line 125 of env.go

or4cle <notifications@github.com> ezt írta (időpont: 2017. jan. 18., Sze,
19:21):

**Comment by or4cle** [comment]
Perhaps something is wrong / off with my test setup, but I am getting various test failures in the ora.v4 branch, even without uncommenting that line. Is there a specific test (or set of tests) that I should be focusing on?

For example, leaving the line commented, I still occasionally see a SIGSEGV like this:

```
--- FAIL: TestPool (3.97s)
        z_server_session_test.go:139: After work: (120,66) -> (124,71)
        z_server_session_test.go:141: After work process count went up 8!
        z_server_session_test.go:144: After work session count went up 19!
        z_server_session_test.go:139: Pool close: (124,71) -> (125,71)
fatal error: unexpected signal during runtime execution
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x381c209470]

runtime stack:
runtime.throw(0x91bf44, 0x2a)
        /usr/local/go/src/runtime/panic.go:566 +0x95
runtime.sigpanic()
        /usr/local/go/src/runtime/sigpanic_unix.go:12 +0x2cc

goroutine 1 [syscall, locked to thread]:
runtime.cgocall(0x82d2b0, 0xc431bf1158, 0x0)
        /usr/local/go/src/runtime/cgocall.go:131 +0x110 fp=0xc431bf1108 sp=0xc431bf10c8
gopkg.in/rana/ora%2ev4._Cfunc_OCIStmtPrepare2(0x7fc06c057120, 0xc4280da000, 0x113e4b8, 0x7fc0d0025c30, 0x43, 0x0, 0x100000000, 0x0, 0x0)
        ??:0 +0x50 fp=0xc431bf1158 sp=0xc431bf1108
gopkg.in/rana/ora%2ev4.(*Ses).Prep(0xc431b360b0, 0xc4285080a0, 0x43, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)

[ output truncated ]
```

**Comment by tgulacsi** [comment]
go test -run=Open

or4cle <notifications@github.com> ezt írta (időpont: 2017. jan. 18., Sze,
21:32):

**Comment by or4cle** [comment]
Thanks for that--I was able to reproduce. It looks like we can avoid it if we deallocate the error handle first, and then the env handle. In this order, it passes on both v3 and v4. This also aligns to some of the sample code from the documentation, e.g.: [here](http://docs.oracle.com/database/121/ADFNS/adfns_cqn.htm#BAJHEGDB).

I'll submit another push request in a moment.

**Comment by tgulacsi** [comment]
I've merged it and released as v3.8.3

Thanks for all!
